# Main entrypoint of the workflow. 
# Please follow the best practices: 
# https://snakemake.readthedocs.io/en/stable/snakefiles/best_practices.html,
# in particular regarding the standardized folder structure mentioned there. 

from snakemake.utils import min_version
from pathlib import Path
import pandas as pd

min_version("6.0")

configfile: "config/config.yaml"
project_name = config["project_name"]
project_dir = Path("results/" + config["project_name"])
input_dir = Path(config["raw_data"])
genome_reference = config["genome_reference"]
output_dir = Path(config["outputfolder"])
gff3_file = config["pombe_gff"]
peptide_stats = config["pombe_peptide"]
fai_file = config["pombe_fai"]
gene_product_file = config["gene_product"]
gene_essentiality_file = config["gene_essentiality"]
output_dir = Path(config["outputfolder"])

samples_with_timepoints = pd.read_csv(config["samples_with_timepoints"], index_col=0)
samples = samples_with_timepoints.columns.tolist()
timepoints = samples_with_timepoints.index.tolist()

rule all:
    input:
        output_dir / "nucletides_for_each_coding_region.bed",
        expand(project_dir/"7_concated/{sample}/{sample}.Reads", sample=samples),
        # project_dir/f"9_annotated_insertions/{config['project_name']}.annotated.csv"
        # expand(project_dir/"12_FWHM_filtered_reads/{sample}.normalized.csv", sample=samples),
        # expand(project_dir/"6_merged/{sample}/{sample}-{timepoint}.bed", sample=samples, timepoint=timepoints),
        # directory(project_dir/"11_normalization/")
        # expand(project_dir/f"15_curve_fitting/{project_name}.{{smooth}}.site.csv", smooth=["before_smoothing", "after_smoothing"]),
        # expand(project_dir/f"15_curve_fitting/{project_name}.{{smooth}}.gene.csv", smooth=["before_smoothing", "after_smoothing"])
        # project_dir/f"14_integrated_values/{project_name}.GMs.csv",
        # project_dir/f"14_integrated_values/{project_name}.annotated_with_domains.csv"
        # project_dir/f"15_curve_fitting/{project_name}.LOWESSTrue.domain.csv"
        # expand(project_dir/"12_FWHM_filtered_reads/{sample}.A.csv", sample=samples),
        # expand(project_dir/f"15_curve_fitting/{project_name}.fit{{fitting}}.pred{{prediction}}.LOWESS{{LOWESS}}.fatol{{fatol}}.{{level}}.csv", fitting=["points5"], prediction=["points5"], LOWESS=[0, 1], level=["domain", "gene"], fatol=config["fatols"])


include: "rules/making_genome_intervals.smk"
include: "rules/preprocessing.smk"
include: "rules/annotating_and_normalizing.smk"
include: "rules/depletion_curve_smoothing.smk"
