{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "423db1c9",
   "metadata": {},
   "source": [
    "# 1. Import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "07bc1ea6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "from pathlib import Path\n",
    "from tqdm import tqdm\n",
    "\n",
    "# scientific\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# bio\n",
    "import gffutils\n",
    "\n",
    "# custom\n",
    "sys.path.append(str(Path.cwd().parent / \"src\"))\n",
    "from utils import update_sysIDs\n",
    "from enrichment_functions import OntologyDataConfig, load_ontology_data\n",
    "from pombe_feature_functions import config, DNA_level_features, extract_protein_features_from_peptide_sequence\n",
    "from protein_structure_functions import pLDDT_statistics_report"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6a46f05",
   "metadata": {},
   "source": [
    "# 2. Constants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d7320181",
   "metadata": {},
   "outputs": [],
   "source": [
    "ALPHAFOLD_DATABASE_DIR = Path(\"../../../resource/AlphaFold_Dataset/20251107_downloaded/UP000002485_284812_SCHPO_v6\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06f567c5",
   "metadata": {},
   "source": [
    "# 3. Configurations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2c9b0058",
   "metadata": {},
   "outputs": [],
   "source": [
    "cfg = config()\n",
    "db = gffutils.create_db(cfg.gff3_file, cfg.database_file, force=True)\n",
    "db = gffutils.FeatureDB(cfg.database_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2ae6a9d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta = pd.read_csv(cfg.gene_meta_file, sep=\"\\t\")\n",
    "uniprot2id = dict(zip(gene_meta[\"uniprot_id\"], gene_meta[\"gene_systematic_id\"]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbc5634a",
   "metadata": {},
   "source": [
    "# 4. Collect features"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f1e6f82",
   "metadata": {},
   "source": [
    "## 4.1 DNA-level features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a0aa1b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "mRNAs = list(db.features_of_type('mRNA'))\n",
    "coding_gene_DNA_features = []\n",
    "\n",
    "for ix, mRNA in enumerate(mRNAs):\n",
    "    gene_DNA_features = DNA_level_features.from_gffutils_feature(mRNA, db, cfg)\n",
    "    coding_gene_DNA_features.append(gene_DNA_features)\n",
    "\n",
    "coding_gene_DNA_features_df = pd.DataFrame(coding_gene_DNA_features)\n",
    "coding_genes = coding_gene_DNA_features_df['Gene_id'].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6586ff2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "coding_gene_DNA_features_df.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c844712",
   "metadata": {},
   "outputs": [],
   "source": [
    "coding_gene_DNA_features_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b01c488",
   "metadata": {},
   "outputs": [],
   "source": [
    "coding_gene_DNA_features_df.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2cdf20e",
   "metadata": {},
   "source": [
    "## 4.2 RNA-level features"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6361b524",
   "metadata": {},
   "source": [
    "### RNA abundance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f47a1c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "mRNA_abundance = pd.read_excel(\"../../resources/Literature/PMID23101633/mmc1.xlsx\", sheet_name=\"Table_S2\", comment=\"#\").set_index('Systematic.name')\n",
    "mRNA_abundance = mRNA_abundance[['MM1.tot.cpc_ex', 'MM2.tot.cpc_ex', 'MN1.tot.cpc_ex', 'MN2.tot.cpc_ex']].copy()\n",
    "mRNA_abundance.columns = pd.MultiIndex.from_tuples([(\"EMM_Proliferating_Cell_RNA_Abundance\", \"replicate1\"), (\"EMM_Proliferating_Cell_RNA_Abundance\", \"replicate2\"), (\"EMM_Nitrogen_Starved_Cell_RNA_Abundance\", \"replicate1\"), (\"EMM_Nitrogen_Starved_Cell_RNA_Abundance\", \"replicate2\")], name=[\"Condition\", \"Replicate\"])\n",
    "mean_mRNA_abundance = mRNA_abundance.T.groupby(level=\"Condition\").mean().T\n",
    "std_mRNA_abundance = mRNA_abundance.T.groupby(level=\"Condition\").std().T\n",
    "cv_mRNA_abundance = std_mRNA_abundance / mean_mRNA_abundance\n",
    "mRNA_abundance_statistics = pd.concat([mean_mRNA_abundance, std_mRNA_abundance, cv_mRNA_abundance], axis=1, keys=['mean', 'std', 'cv'])\n",
    "mRNA_abundance_statistics.index = update_sysIDs(mRNA_abundance_statistics.index.tolist(), cfg.gene_meta_file)\n",
    "coding_gene_mRNA_abundace_statistics = mRNA_abundance_statistics[mRNA_abundance_statistics.index.isin(coding_genes)].copy().dropna().round(3)\n",
    "coding_gene_mRNA_abundace_statistics.columns = [\"_\".join(col).strip() for col in coding_gene_mRNA_abundace_statistics.columns.values]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d471541",
   "metadata": {},
   "source": [
    "### RNA kinetics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1772ccf",
   "metadata": {},
   "outputs": [],
   "source": [
    "mRNA_kinetics = pd.read_excel(\"../../resources/Literature/PMID27825301/12864_2016_3237_MOESM22_ESM.xls\", sheet_name=\"Table\")\n",
    "mRNA_kinetics = mRNA_kinetics[[\"Gene ID\", \"tAIg\", \"HL - Mata (5)\", \"SR - Mata (5)\"]].set_index(\"Gene ID\")\n",
    "mRNA_kinetics.columns = [\"tAIg\", \"mRNA_half_life_minutes\", \"mRNA_synthesis_rate_per_minute\"]\n",
    "mRNA_kinetics.index = update_sysIDs(mRNA_kinetics.index.tolist(), cfg.gene_meta_file)\n",
    "coding_gene_mRNA_kinetics = mRNA_kinetics[mRNA_kinetics.index.isin(coding_genes)].copy().dropna().round(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbad2ef6",
   "metadata": {},
   "source": [
    "### Concatenate RNA features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10e0efd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "coding_gene_RNA_features_df = pd.merge(coding_gene_mRNA_abundace_statistics, coding_gene_mRNA_kinetics, left_index=True, right_index=True, how=\"outer\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d24b3d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "coding_gene_RNA_features_df.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad17b3a1",
   "metadata": {},
   "source": [
    "## 4.3 Protein-level features"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "232ae47b",
   "metadata": {},
   "source": [
    "### Protein basic properties"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e773e419",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_metadata = cfg.protein_meta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "044e2152",
   "metadata": {},
   "source": [
    "### Protein features from the peptide sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a406d085",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/a/yangyusheng/miniforge3/envs/bioinformatics/lib/python3.12/site-packages/Bio/SeqUtils/ProtParam.py:106: BiopythonDeprecationWarning: The get_amino_acids_percent method has been deprecated and will likely be removed from Biopython in the near future. Please use the amino_acids_percent attribute instead.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Stop codon found in sequence of SPAC2E12.05. Truncating at first stop codon.\n"
     ]
    }
   ],
   "source": [
    "protein_features_from_peptide_sequences = extract_protein_features_from_peptide_sequence(cfg.PomBase_resource_dir / \"genome_sequence_and_features\" / \"peptide.fa\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b92e6f70",
   "metadata": {},
   "source": [
    "### Protein abundance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7108db85",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_abundance = pd.read_csv(cfg.PomBase_resource_dir / \"RNA_metadata\" / \"quantitative_gene_expression.tsv\", sep=\"\\t\")\n",
    "EMM_proliferating_cell_protein_abundance = gene_abundance.query(\"reference == 'PMID:23101633' and type == 'protein' and condition == 'glucose MM,standard temperature'\")[[\"gene_systematic_id\", \"copies_per_cell\"]].dropna().astype({\"copies_per_cell\": float})\n",
    "EMMN_quiescent_cell_protein_abundance = gene_abundance.query(\"reference == 'PMID:23101633' and type == 'protein' and condition == 'glucose MM,nitrogen absent,standard temperature'\")[[\"gene_systematic_id\", \"copies_per_cell\"]].dropna().astype({\"copies_per_cell\": float})\n",
    "protein_abundance = pd.merge(EMM_proliferating_cell_protein_abundance, EMMN_quiescent_cell_protein_abundance, left_on=\"gene_systematic_id\", right_on=\"gene_systematic_id\", suffixes=(\"_EMM_Proliferating_Cell\", \"_EMMN_Quiescent_Cell\")).set_index(\"gene_systematic_id\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "472d7e29",
   "metadata": {},
   "source": [
    "### Protein turnover"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7adef5b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_kinetics = pd.read_excel(\"../../resources/Literature/PMID25466257/1-s2.0-S2211124714009346-mmc3.xlsx\", na_values=[\"n.d.\"]).dropna(subset=[\"Degradation rates (min-1)\", \"t1/2 (min)\"])\n",
    "protein_kinetics[\"ENSG\"] = protein_kinetics[\"ENSG\"].fillna(protein_kinetics[\"Gene name\"])\n",
    "protein_kinetics = protein_kinetics[[\"ENSG\", \"t1/2 (min)\"]].set_index(\"t1/2 (min)\")\n",
    "protein_kinetics = protein_kinetics[\"ENSG\"].str.split(\";\").explode().reset_index()\n",
    "protein_kinetics[\"gene_systeamtic_id\"] = update_sysIDs(protein_kinetics[\"ENSG\"].tolist(), cfg.gene_meta_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "241b7444",
   "metadata": {},
   "source": [
    "### pLDDT features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28c1ee59",
   "metadata": {},
   "outputs": [],
   "source": [
    "pLDDTs = pLDDT_statistics_report(\n",
    "    structure_dir=ALPHAFOLD_DATABASE_DIR,\n",
    "    structure_format=\"pdb.gz\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c086456c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pLDDTs[\"Systematic_ID\"] = pLDDTs[\"uniprot_id\"].map(uniprot2id)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bdb051c",
   "metadata": {},
   "source": [
    "### Domain features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ff5cad48",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_domains = pd.read_csv(cfg.PomBase_resource_dir / \"Protein_features\" / \"protein_families_and_domains.tsv\", sep=\"\\t\")\n",
    "pfam_domain_counts = protein_domains.query(\"database == 'PFAM'\").groupby(\"systematic_id\").size().rename(\"PFAM_domain_count\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8553cbf",
   "metadata": {},
   "source": [
    "### Concatenate protein features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "132d78e1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "0b625d79",
   "metadata": {},
   "source": [
    "## 4.4 Evolutionary-level features"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32e29fbd",
   "metadata": {},
   "source": [
    "### Ortholog count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "988b7ed2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ortholog_counts(ortholog_file: Path) -> pd.Series:\n",
    "    \"\"\"Get the number of orthologs for each gene from the ortholog file.\"\"\"\n",
    "    ortholog_df = pd.read_csv(ortholog_file, sep=\"\\t\", index_col=0, header=None, names=[\"gene_systematic_id\", \"orthologs\"], na_values=\"NONE\")\n",
    "    ortholog_counts = ortholog_df[\"orthologs\"].str.split(\"|\").apply(lambda x: len(x) if isinstance(x, list) else 0 )\n",
    "    return ortholog_counts\n",
    "\n",
    "num_of_orthologs_with_japonicus = get_ortholog_counts(cfg.PomBase_resource_dir / \"curated_orthologs\" / \"pombe_japonicus_orthologs.txt\")\n",
    "num_of_orthologs_with_cerevisiae = get_ortholog_counts(cfg.PomBase_resource_dir / \"curated_orthologs\" / \"pombe_cerevisiae_orthologs.txt\")\n",
    "num_of_orthologs_with_human = get_ortholog_counts(cfg.PomBase_resource_dir / \"curated_orthologs\" / \"pombe_human_orthologs.txt\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0742a951",
   "metadata": {},
   "source": [
    "### Paralog count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39e4ca5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "pombe_paralogs = pd.read_csv(\"../../resources/pombe_paralog_from_ensemble_biomart_export.tsv\", sep=\"\\t\")\n",
    "\n",
    "paralog_count = pombe_paralogs.query(\"`Gene stable ID` in @coding_genes\").groupby([\"Gene stable ID\", \"Gene name\"]).apply(lambda sub_df: sub_df.shape[0], include_groups=False).rename(\"paralog_count\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "796fa26d",
   "metadata": {},
   "source": [
    "## 4.5 Network-level features"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13b17004",
   "metadata": {},
   "source": [
    "### GO term richness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b7464c3a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/c/yangyusheng_optimized/DIT_HAP_pipeline/resources/pombase_data/2025-10-01/ontologies_and_associations/go-basic.obo: fmt(1.2) rel(2025-10-10) 42,666 Terms; optional_attrs(def relationship)\n",
      "HMS:0:00:01.890109  49,447 annotations READ: /data/c/yangyusheng_optimized/DIT_HAP_pipeline/resources/pombase_data/2025-10-01/ontologies_and_associations/gene_ontology_annotation.gaf.tsv \n"
     ]
    }
   ],
   "source": [
    "dag, objanno, ns2assoc, gene2go, go2genes, slim_dag, ns2slim_assoc = load_ontology_data(\n",
    "    OntologyDataConfig(\n",
    "        ontology_obo=cfg.PomBase_resource_dir / \"ontologies_and_associations\" / \"go-basic.obo\",\n",
    "        ontology_association_gaf=cfg.PomBase_resource_dir / \"ontologies_and_associations\" / \"gene_ontology_annotation.gaf.tsv\",\n",
    "        slim_terms_table=[\n",
    "            cfg.PomBase_resource_dir / \"ontologies_and_associations\" / \"bp_go_slim_terms.tsv\",\n",
    "            cfg.PomBase_resource_dir / \"ontologies_and_associations\" / \"mf_go_slim_terms.tsv\",\n",
    "            cfg.PomBase_resource_dir / \"ontologies_and_associations\" / \"cc_go_slim_terms.tsv\"\n",
    "        ]\n",
    "    ).load_data(),\n",
    "    **{\n",
    "        \"relationships\": {\"is_a\", \"part_of\"},\n",
    "        \"propagate_counts\": True,\n",
    "        \"load_obsolete\": False,\n",
    "        \"prt\": None\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "844dfdcd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Calculating GO term richness: 100%|██████████| 5262/5262 [00:00<00:00, 190203.28it/s]\n"
     ]
    }
   ],
   "source": [
    "gene2gorichness = {}\n",
    "for gene, go_terms in tqdm(gene2go.items(), desc=\"Calculating GO term richness\"):\n",
    "    gene2gorichness[gene] = len(set(go_terms))\n",
    "\n",
    "go_richness = pd.DataFrame.from_dict(gene2gorichness, orient=\"index\", columns=[\"GO_term_richness\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f5778f6",
   "metadata": {},
   "source": [
    "### PPI degree and GI degree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5770aec2",
   "metadata": {},
   "outputs": [],
   "source": [
    "biogrid_data = pd.read_csv(\"../../resources/BioGrid/BIOGRID-ORGANISM-Schizosaccharomyces_pombe_972h-5.0.251.tab3.txt\", sep=\"\\t\")\n",
    "PPI_and_GI = biogrid_data[['Systematic Name Interactor A',\n",
    "       'Systematic Name Interactor B', 'Official Symbol Interactor A',\n",
    "       'Official Symbol Interactor B', 'Experimental System Type']].drop_duplicates()\n",
    "\n",
    "PPI = PPI_and_GI.query(\"`Experimental System Type` == 'physical'\")\n",
    "GI = PPI_and_GI.query(\"`Experimental System Type` == 'genetic'\")\n",
    "\n",
    "PPI_degrees = PPI.groupby('Systematic Name Interactor A').size().rename(\"PPI_degree\")\n",
    "GI_degrees = GI.groupby('Systematic Name Interactor A').size().rename(\"GI_degree\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6562e79",
   "metadata": {},
   "source": [
    "## 4.6 Phenotype-level features"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9883149",
   "metadata": {},
   "source": [
    "### FYPO Essentiality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "76e056ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "FYPO_viability = pd.read_csv(cfg.PomBase_resource_dir / \"Gene_metadata\" / \"gene_viability.tsv\", sep=\"\\t\", header=None, names=[\"gene_systematic_id\", \"FYPOviability\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d963f431",
   "metadata": {},
   "source": [
    "### Deletion Library Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "62681916",
   "metadata": {},
   "outputs": [],
   "source": [
    "DeletionLibrary_essentiality = pd.read_excel(\"../../resources/Hayles_2013_OB_merged_categories_sysIDupdated.xlsx\")[[\"Updated_Systematic_ID\", \"Gene dispensability. This study\", \"Category\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "569f7956",
   "metadata": {},
   "source": [
    "### Transposon insertion density"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51708005",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioinformatics",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
